### PR TITLE
Add `weight` to harmonic_diameter

### LINKED ISSUE #7624: Allow edge weights for `harmonic_diameter`
The harmonic diameter (which is the harmonic mean of all distances) from #5251 uses path lengths for unweighted networks. It would be good to allow weights path lengths via a keyword argument like `weight=None`. This kwarg can have the same name doc_string entry and optional function interface as for the shortest_path functions.

### PR DESCRIPTION
**Pull request by Raj3110**
Closes #7624,
 

1. Added a weight parameter to the function signature:

   `pythonCopydef harmonic_diameter(G, weight=None, sp=None):`

2. Updated the docstring to include information about the weight parameter.

3. Modified the shortest path calculation to use nx.single_source_dijkstra_path_length instead of nx.single_source_shortest_path_length:
`pythonCopylength = nx.single_source_dijkstra_path_length(G, n, weight=weight)`

4.Updated the @nx._dispatchable decorator to include the weight parameter:
   `pythonCopy@nx._dispatchable(edge_attrs="weight")`

### CODE REVIEW

File: networkx/algorithms/distance_measures.py [hunk 1]
```diff
@@ -387,8 +387,11 @@ def diameter(G, e=None, usebounds=False, weight=None):
     return max(e.values())
 
 
-@nx._dispatchable
-def harmonic_diameter(G, sp=None):
+import math
+import networkx as nx
+
```
Comment (rossbar) [inline networkx/algorithms/distance_measures.py:392]: This should be removed

File: networkx/algorithms/distance_measures.py [hunk 2]
```diff
@@ -387,8 +387,11 @@ def diameter(G, e=None, usebounds=False, weight=None):
...
+@nx._dispatchable(edge_attrs="weight")
+def harmonic_diameter(G, weight=None, sp=None):
```
Comment (rossbar) [inline networkx/algorithms/distance_measures.py:394]: Generally speaking, switching the order of non-keyword-only arguments is a no-go as it breaks backward compatibility. However, in this case I believe `harmonic_diameter` is not yet included in a released version so it should be fine.

I'd vote to eliminate this problem going forward by enforcing keyword-only arguments:

```suggestion
def harmonic_diameter(G, *, weight=None, sp=None):
```

File: networkx/algorithms/distance_measures.py [hunk 3]
```diff
@@ -412,6 +415,10 @@ def harmonic_diameter(G, sp=None):
     G : NetworkX graph
        A graph
 
+    weight : string or None, optional (default=None)
+       The edge attribute that holds the numerical value used as a weight.
+       If None, then each edge has unit weight.
+
```
Comment (rossbar) [inline networkx/algorithms/distance_measures.py:421]: As noted in the linked issue, this should match the parameter description for `single_source_dijkstra_path_length`

File: networkx/algorithms/distance_measures.py [hunk 4]
```diff
@@ -452,7 +459,6 @@ def harmonic_diameter(G, sp=None):
         return math.inf
     return math.nan
 
-
```
Comment (rossbar) [inline networkx/algorithms/distance_measures.py:455]: This has broken the formatter - see the [contributing guide](https://networkx.org/documentation/latest/developer/contribute.html) for info on how to automatically conform to the project's formatting rules.

**Review by rossbar — CHANGES_REQUESTED** [review]
This seems like the right idea, but some details still need some work.

Also - is there a reason to choose one weighted shortest path algorithm (`dijkstra`) over another (e.g. `bellman_ford`)?

**Comment by Raj3110** [comment]
> This seems like the right idea, but some details still need some work.
> 
> Also - is there a reason to choose one weighted shortest path algorithm (`dijkstra`) over another (e.g. `bellman_ford`)?

I preferred Dijkstra's algorithm over Bellman-Ford for weighted shortest paths due to its better performance, especially in sparse graphs with non-negative weights. However, Bellman-Ford is necessary for graphs with negative weights or to detect negative cycles. Here in NetworkX we used Dijkstra's in functions like `eccentricity` and `barycenter`, assuming non-negative weights for general-purpose graph analysis.

File: networkx/algorithms/distance_measures.py [hunk 5]
```diff
@@ -387,52 +387,72 @@ def diameter(G, e=None, usebounds=False, weight=None):
...
-    infinite distances (e.g., a perfect match has harmonic diameter equal
-    to the number of vertices minus one), making it possible to assign a
-    meaningful value to all graphs.
-
-    Note that in [1] the harmonic diameter is called "connectivity length":
-    however, "harmonic diameter" is a more standard name from the
-    theory of metric spaces. The name "harmonic mean distance" is perhaps
-    a more descriptive name, but is not used in the literature, so we use the
-    name "harmonic diameter" here.
+       The harmonic diameter of a graph is the harmonic mean of the distances
```
Comment (rossbar) [inline networkx/algorithms/distance_measures.py:394]: The changes to indentation level should be undone - they are not stylistically correct and make reviewing much more difficult!

File: networkx/algorithms/distance_measures.py [hunk 6]
```diff
@@ -412,6 +412,26 @@ def harmonic_diameter(G, sp=None):
...
+    positional arguments: the two endpoints of an edge and the
+    dictionary of edge attributes for that edge. The function must
+    return a number.
+
+    If this is None, every edge has weight/distance/cost 1.
+
+    Weights stored as floating point values can lead to small round-off
+    errors in distances. Use integer weights to avoid this.
+
+    Weights should be positive, since they are distances.
```
Comment (dschult) [inline networkx/algorithms/distance_measures.py:433]: Can you use the doc_string language for `weight` from one of the existing functoins that allows `weight` to be a function? That provides consistency and eases review.

Comment (Raj3110) [reply networkx/algorithms/distance_measures.py]: 1. Updated the `weight` parameter description to match the style used in other NetworkX functions.

2. Moved the notes about floating-point weights and positive weights to a separate "Notes" section at the end of the docstring.

3. Kept the rest of the docstring, including the detailed explanation and references, unchanged.

**Review by rossbar — APPROVED** [review]
Thanks for the multiple rounds @Raj3110 , this LGTM now!

**Comment by rossbar** [comment]
Another improvement would be to add tests to ensure the weighted `harmonic_diameter` is correct!

**Comment by rossbar** [comment]
Since this didn't make the 3.4 release, I took the liberty of pushing up a signature change to maintain backward compatibility. My approval stands, though testing would be welcome!

**Comment by dschult** [comment]
I took the liberty of pushing up some tests.

I think this is now ready. @rossbar can you take a look at the tests?

**Comment by rossbar** [comment]
> I think this is now ready. @rossbar can you take a look at the tests?

Tests LGTM, thanks @dschult !

**Review by dschult — APPROVED** [review]

**Pull request merged**